ddspipe_core/src/cpp/efficiency/payload/PayloadPoolMediator.cpp

```diff
@@ -0,0 +1,170 @@
+        types::RtpsPayloadData* data)
+{
+    // Lock the mutex_ to make sure that when we get the PayloadPool get_payload call,
+    // we haven't replaced the data.
+    std::lock_guard<std::mutex> lock(mutex_);
+
+    data_ = data;
+
+    return writer->write(data);
+}
+
+bool PayloadPoolMediator::write(
+        fastdds::dds::DataWriter* writer,
+        types::RtpsPayloadData* data,
+        fastrtps::rtps::WriteParams& params)
+{
+    // Lock the mutex_ to make sure that when we get the PayloadPool get_payload call,
+    // we haven't replaced the data.
+    std::lock_guard<std::mutex> lock(mutex_);
+
+    data_ = data;
+
+    return writer->write(data, params);
+}
+
+utils::ReturnCode PayloadPoolMediator::write(
+        fastdds::dds::DataWriter* writer,
+        types::RtpsPayloadData* data,
+        const fastrtps::rtps::InstanceHandle_t& handle)
+{
+    // Lock the mutex_ to make sure that when we get the PayloadPool get_payload call,
+    // we haven't replaced the data.
+    std::lock_guard<std::mutex> lock(mutex_);
+
+    data_ = data;
+
+    return writer->write(data, handle);
+}
+
+bool PayloadPoolMediator::get_payload(
+        uint32_t size,
+        fastrtps::rtps::CacheChange_t& cache_change)
+{
+    fastrtps::rtps::IPayloadPool* payload_owner =
+            dynamic_cast<IPayloadPool*>(payload_pool_.get());
+
+    if (get_payload(data_->payload, payload_owner, cache_change))
+    {
+        // Replace the pointer to destroy the data.
+        data_ = nullptr;
+
+        cache_change.payload_owner(payload_pool_.get());
+        return true;
+    }
+    else
+    {
+        logDevError(DDSPIPE_PAYLOADPOOL, "Error occurred while creating payload.")
+        return false;
+    }
+}
+
+bool PayloadPoolMediator::get_payload(
+        fastrtps::rtps::SerializedPayload_t& data,
+        fastrtps::rtps::IPayloadPool*& data_owner,
+        fastrtps::rtps::CacheChange_t& cache_change)
+{
+    if (payload_pool_->get_payload(data, data_owner, cache_change.serializedPayload))
+    {
+        cache_change.payload_owner(payload_pool_.get());
+        return true;
+    }
+    else
+    {
+        logDevError(DDSPIPE_PAYLOADPOOL, "Error occurred while referencing payload.")
+        return false;
+    }
+}
+
+bool PayloadPoolMediator::release_payload(
+        fastrtps::rtps::CacheChange_t& cache_change)
+{
+    if (cache_change.payload_owner() == payload_pool_.get())
+    {
+        if (payload_pool_->release_payload(cache_change.serializedPayload))
+        {
+            cache_change.payload_owner(nullptr);
+            return true;
+        }
+        else
+        {
+            logDevError(DDSPIPE_PAYLOADPOOL, "Error occurred while releasing payload.")
+            return false;
+        }
+    }
+    else
+    {
+        logError(DDSPIPE_PAYLOADPOOL, "Trying to remove a cache change in an incorrect pool.")
+        throw utils::InconsistencyException("Trying to remove a cache change in an incorrect pool.");
+    }
```

Comment:
I wouldn't say the mediator should have this logic, but the pool it manages. Same for other methods.